### ISSUE TITLE
Document only a library's exported files from a given starting point

### ISSUE THREAD

**Issue by manrajgrover** [TypeStrong/typedoc#639]
Currently, `typedoc` exports all internally exported entities which are not part of the exports in the given starting point. I've been trying to find a way to get past that using available options but looks like this feature is not available.

For example, if starting point is `src/index.ts`, then only exports from this file should be documented and not all entities. Restricting files in config file or marking every internal utility function as `@hidden` makes things unmaintainable.

If I have missed anything, please let me know.

**Comment by ManrajGrover** [comment]
@aciccarello Any updates on this?

**Comment by aciccarello** [comment]
@ManrajGrover No movement yet. I like the idea though. Currently the "interval" vs "external" concepts are pretty limited and probably need to be thought through. There is a [community made plugin](https://github.com/christopherthielen/typedoc-plugin-internal-external) to override these via doc tags if you are interested but it may not fit the bill of what you are looking for.

PRs are welcome :heart:

**Comment by ManrajGrover** [comment]
@aciccarello Thanks for the update. I'm pretty new to TypeScript Compiler API and although I tried going through their examples, I found their documentation lacking in this regard. I would appreciate if you could share more information/pointers on the same.

**Comment by aciccarello** [comment]
@ManrajGrover I didn't write TypeDoc so I don't have much experience with the compiler API. I can tell you that TypeDoc uses some [internal TypeScript apis](https://github.com/TypeStrong/typedoc/blob/34377dc2cd7e901655423d2f0039e11bb73f67d2/src/lib/ts-internal.ts) that aren't included in the public api.

**Comment by aciccarello** [comment]
I'm thinking this would be a new mode (maybe named "library"). I not sure how to handle projects with multiple entry points (e.g. `rxjs/Observable` vs `rxjs/Subject`). Ideas are welcome.

**Comment by franklinyu** [comment]
Should the new "library" mode ignore everything in `node_modules`?

**Comment by aciccarello** [comment]
@franklinyu I don't think that should be conflated with a "library" mode. A library may want to include types for an imported module. Excluding `node_modules` can be discussed further in #319.

**Comment by Gerrit0** [comment]
> I not sure how to handle projects with multiple entry points (e.g. `rxjs/Observable` vs `rxjs/Subject`)

I think this should be handled in basically the same way as we already handle accepting files. `typedoc --mode library rxjs/Observable.ts rxjs/Subject.ts`

**Comment by franklinyu** [comment]
What if

```bash
typedoc --mode library rxjs/*.ts
```

? Regard all those TypeScript sources as entry points?

**Comment by Gerrit0** [comment]
Yes, the `expandInputFiles` method on the Application class will already do this.

**Comment by yordis** [comment]
@Gerrit0 I tried to implement this but I got nowhere. 

I tried to find where in the source code I would filter what gets generated in the documentation as `@hidden` do it.

Could you give me some ideas of the source code so I could try to fix this? Too much code right now so any extra help will be appreciated.

**Comment by Gerrit0** [comment]
I think the best place to filter would probably be when first generating reflections, so no reflection would be generated if it doesn't come from an input file...

It seems like the place to do this would be [createDeclaration](https://github.com/TypeStrong/typedoc/blob/a2fab7cf1b5957c2e05ffd3882583dcf812ea8c7/src/lib/converter/factories/declaration.ts#L35), since other filters (excludeNotExported) are applied there.

I'm not sure it's the best place, but it seems to be the right place for now.

**Comment by yordis** [comment]
Alright around here https://github.com/TypeStrong/typedoc/blob/a2fab7cf1b5957c2e05ffd3882583dcf812ea8c7/src/lib/converter/factories/declaration.ts#L53

I want to exit if the reflection is not being exported and the reflection is an actual module export.

I am having a little bit of trouble understanding https://github.com/TypeStrong/typedoc/blob/1c4c5151cec938ef03cae79c33c12be8d587ea74/src/lib/models/reflections/abstract.ts#L32-L66 types and what they actually mean.

And these https://github.com/TypeStrong/typedoc/blob/1c4c5151cec938ef03cae79c33c12be8d587ea74/src/lib/models/reflections/abstract.ts#L68-L84

I see some keys like these https://github.com/TypeStrong/typedoc/blob/1c4c5151cec938ef03cae79c33c12be8d587ea74/src/lib/models/reflections/abstract.ts#L74-L75 but I am not quite sure what they are.

Right now I need to figure out which types are useful.

I don't care if the thing is a variable declaration, function, or whatever, as long as it is actually being exported.

I tried to use https://github.com/TypeStrong/typedoc/blob/1c4c5151cec938ef03cae79c33c12be8d587ea74/src/lib/models/reflections/abstract.ts#L140

But even interface keys will be `true` it wouldn't work to use the key by itself since the filtering is for things being directly exported on the module.


Sorry for the confusion, I am trying to put the pieces together.

**Comment by Gerrit0** [comment]
Yea, that `isExported` on the reflections class won't be helpful since you are trying to determine exports from the entry point.

I think the way to do this is to get the module symbol from the entry point, and then see if a given node's symbol is in the exported map of that module. You'll need to access TS apis as this knowledge isn't included in the typedoc reflections yet.

**Comment by yordis** [comment]
@Gerrit0 any extra directions in terms of file and where to start looking for this will be helpful. I will give another try to this so I can find that information from the TS.

**Comment by Geo1088** [comment]
Any movement on this? If that code works, can it be PR'd? I'm itching to use this to document a library I'm writing.

**Comment by capraynor** [comment]
> [shrinktofit@ab2f239](https://github.com/shrinktofit/typedoc/commit/ab2f239c8e9710445228effada2640c863ece997)
> Please take a look at this commit in my fork reposity.
> It support:
> 
> ```
> export * from "xx";
> export { A, B as C } from "yy";
> 
> import * as D from "zz";
> export { D };
> ```
> 
> If your project is made as a single library so that it can be imported by user from a root `index.ts`(or a bundle packed by packing-tools), the result will exactly contain the entities that your `index.ts` exported.
> At this point, only one external module exists. (If you call typedoc with only one input file).
> 
> ![image](https://user-images.githubusercontent.com/28807867/53147314-2c158680-35e2-11e9-8edc-fc7d740bff5e.png)
> 
> ![image](https://user-images.githubusercontent.com/28807867/53147323-33d52b00-35e2-11e9-81db-bc99c9ae12b3.png)

Man this commit is so cool. Solved my problem.

**Comment by FranklinYu** [comment]
@shrinktofit Ready for PR? I think this should be the solution.

**Comment by englercj** [comment]
The referenced commit no longer seems to be available, is there a different workaround for documenting node modules that only export a subset of the objects in the modules?

Would love to specify only my `index.ts` file and let typedoc crawl the exports.

**Comment by Geo1088** [comment]
Still available for me, https://github.com/shrinktofit/typedoc/tree/leslie

It appears I'm able to create a pull request from another user's fork, but I feel like that's in poor taste. Would rather @shrinktofit push this forward themselves.

**Comment by shrinktofit** [comment]
Sorry for the late. The PR is available now.

**Comment by rikakomoe** [comment]
> [shrinktofit@ab2f239](https://github.com/shrinktofit/typedoc/commit/ab2f239c8e9710445228effada2640c863ece997)
> Please take a look at this commit in my fork reposity.
> It support:
> 
> ```
> export * from "xx";
> export { A, B as C } from "yy";
> 
> import * as D from "zz";
> export { D };
> ```
> 
> If your project is made as a single library so that it can be imported by user from a root `index.ts`(or a bundle packed by packing-tools), the result will exactly contain the entities that your `index.ts` exported.
> At this point, only one external module exists. (If you call typedoc with only one input file).
> 
> ![image](https://user-images.githubusercontent.com/28807867/53147314-2c158680-35e2-11e9-8edc-fc7d740bff5e.png)
> 
> ![image](https://user-images.githubusercontent.com/28807867/53147323-33d52b00-35e2-11e9-81db-bc99c9ae12b3.png)

Is there any way to omit the `index` / `index.d`? That feels bad.

**Comment by Gerrit0** [comment]
With the work done for export declarations, I think it is almost possible to finally do this, once one of the implementations is merged, I'll give developing a library mode a shot.

**Comment by Gerrit0** [comment]
So it turns out that actually implementing library mode was way easier than I anticipated.

If you wanted this feature, please check out #1184. There is a beta release out with the functionality, but I'd like to iron out any kinks before merging to master.

**Comment by y-nk** [comment]
@Gerrit0 this still an open issue, right? library mode isn't there anymore and even for mode file, typedoc is scanning through every variable/function/type declared on a file level, exported or not. Is there a way to reuse the code from this PR and add a flag --onlyExported ?

**Comment by Gerrit0** [comment]
Yes, this is still an open issue. It is being worked, but due to issues discovered with the initial implementation, has grown considerably larger than expected. https://github.com/TypeStrong/typedoc/pull/1184#issuecomment-650809143 is the latest update on it.

**Comment by y-nk** [comment]
@Gerrit0  thank you for the update. my knowledge of the project is still very limited but let me know if i can be of help.

**Comment by Gerrit0** [comment]
Of all the issues for me to miss updating on release...

0.20 has been released, which behaves as described in this issue.

**Comment by sporritt** [comment]
@Gerrit0 I know this issue is closed but I can't figure out whether "library mode" can generate output like that shown by @shrinktofit above:

<img width="872" alt="image" src="https://user-images.githubusercontent.com/262720/124738471-4d316480-df5c-11eb-9c94-1d290a7d6030.png">

specifically the concept of multiple modules. I have a base directory with a bunch of modules:

<img width="278" alt="image" src="https://user-images.githubusercontent.com/262720/124738770-9bdefe80-df5c-11eb-89e4-9f5b823416da.png">

i can library mode on each of these directories individually, for instance:

```
npx typedoc --mode library --out docs/core ./ts/core
```

but then i don't have any link between the modules. I'd like to run

```
npx typedoc --mode library --out docs/core ./ts
```

but then i just get everything in one.

**Comment by sporritt** [comment]
...well in fact i'm not using library mode it turns out, as I'm only on 0.16.11. I can't upgrade to 0.20 though because it needs typescript 4, and the shift to declaring it as a peer dependency means it won't install its own version as 0.16.11 did. But that's another matter. 

Still would be interested to hear if in 0.20 I should be able to point at the `ts` directory shown above and get one module for each of the directories in there.

**Comment by Gerrit0** [comment]
Yes, with 0.20+ you can do that and more. 0.16.11 must have a bug where invalid options don't produce errors... that should be producing an error. 0.20.37 supports TypeScript 3.9 if I recall correctly, so you might still be able to upgrade. 3.9 support wasn't dropped until 0.21.

The command won't be exactly what you specified (unless each folder has only a single `index.ts` file), however.

```bash
npx typedoc --out docs/core ./ts/browser-ui/index.ts ./ts/connector-bezier/index.ts ./ts/connector-flowchart/index.ts ./ts/core/index.ts ./ts/util/index.ts
```

... at this point, I'd probably switch to a config file.

**Comment by sporritt** [comment]
Thanks for the reply - much appreciated.
